File: api/app/schemas/graph_response.py
import datetime
import uuid
from typing import Any, Literal

from pydantic import BaseModel, Field


class GraphEditorGraphV1(BaseModel):
    id: uuid.UUID
    name: str
    node_count: int
    folder_id: uuid.UUID | None = None
    workspace_id: uuid.UUID | None = None
    description: str | None = None
    temporary: bool = False
    pinned: bool = False
    created_at: datetime.datetime | None = None
    updated_at: datetime.datetime | None = None
    custom_instructions: list[str] = Field(default_factory=list)
    max_tokens: int | None = None
    temperature: float | None = None
    top_p: float | None = None
    top_k: int | None = None
    frequency_penalty: float | None = None
    presence_penalty: float | None = None
    repetition_penalty: float | None = None
    reasoning_effort: str | None = None


class GraphEditorNodeV1(BaseModel):
    id: str
    type: str
    position_x: float
    position_y: float
    width: str = "100px"
    height: str = "100px"
    parent_node_id: str | None = None
    data: dict[str, Any] | list[Any] | None = None


class GraphEditorEdgeV1(BaseModel):
    id: str
    source_node_id: str
    target_node_id: str
    source_handle_id: str | None = None
    target_handle_id: str | None = None
    type: str | None = None
    label: str | None = None
    animated: bool = False
    style: dict[str, Any] | list[Any] | None = None
    data: dict[str, Any] | list[Any] | None = None


class GraphEditorResponseV1(BaseModel):
    version: Literal[1]
    graph: GraphEditorGraphV1
    nodes: list[GraphEditorNodeV1]
    edges: list[GraphEditorEdgeV1]

File: ui/app/types/graphResponse.ts
import type { ReasoningEffortEnum } from '@/types/enums';
import type { Graph } from '@/types/graph';

export const GRAPH_EDITOR_RESPONSE_VERSION = 1 as const;

export type GraphJsonContainer = Record<string, JsonValue> | unknown[];

export interface GraphEditorGraphV1 {
    id: string;
    name: string;
    node_count: number;
    folder_id?: string | null;
    workspace_id?: string | null;
    description?: string | null;
    temporary?: boolean;
    pinned?: boolean;
    created_at?: string | null;
    updated_at?: string | null;
    custom_instructions?: string[];
    max_tokens?: number | null;
    temperature?: number | null;
    top_p?: number | null;
    top_k?: number | null;
    frequency_penalty?: number | null;
    presence_penalty?: number | null;
    repetition_penalty?: number | null;
    reasoning_effort?: ReasoningEffortEnum | null;
}

export interface GraphEditorNodeV1 {
    id: string;
    type: string;
    position_x: number;
    position_y: number;
    width?: string;
    height?: string;
    parent_node_id?: string | null;
    data?: GraphJsonContainer | null;
}

export interface GraphEditorEdgeV1 {
    id: string;
    source_node_id: string;
    target_node_id: string;
    source_handle_id?: string | null;
    target_handle_id?: string | null;
    type?: string | null;
    label?: string | null;
    animated?: boolean;
    style?: GraphJsonContainer | null;
    data?: GraphJsonContainer | null;
}

export interface GraphEditorResponseV1 {
    version: typeof GRAPH_EDITOR_RESPONSE_VERSION;
    graph: GraphEditorGraphV1;
    nodes: GraphEditorNodeV1[];
    edges: GraphEditorEdgeV1[];
}

export interface DecodedGraphEditorNodeV1 extends GraphEditorNodeV1 {
    width: string;
    height: string;
    parent_node_id: string | null;
    data: GraphJsonContainer | null;
}

export interface DecodedGraphEditorEdgeV1 extends GraphEditorEdgeV1 {
    source_handle_id: string | null;
    target_handle_id: string | null;
    type: string | null;
    label: string | null;
    animated: boolean;
    style: GraphJsonContainer | null;
    data: GraphJsonContainer | null;
}

export interface DecodedGraphEditorResponse {
    version: typeof GRAPH_EDITOR_RESPONSE_VERSION;
    graph: Graph;
    nodes: DecodedGraphEditorNodeV1[];
    edges: DecodedGraphEditorEdgeV1[];
}
